### PR TITLE
Scala 3 Compilation with 2.12 support

### PR DESCRIPTION
**Pull request by jlprat**
This PR adds support to Scala 3.0 without the need to drop support for Scala 2.12
Did my best as well to add some new stages in Jenkis for Scala 3.0

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jlprat** [comment]
Ping @ijuma I think I managed to keep Scala 2.12 while adding 3.0 support

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala [hunk 1]
```diff
@@ -790,7 +790,7 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
    * @param stateStoreNames   the names of the state store used by the processor
    * @see `org.apache.kafka.streams.kstream.KStream#process`
    */
-  @deprecated(since = "3.0", message = "Use process(ProcessorSupplier, String*) instead.")
+  @deprecated("Use process(ProcessorSupplier, String*) instead.", "3.0")
```
Comment (ijuma) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:793]: Why is this change required?

File: Jenkinsfile [hunk 2]
```diff
@@ -181,6 +181,25 @@ pipeline {
           }
         }
 
+        stage('JDK 11 and Scala 3.0') {
```
Comment (ijuma) [inline Jenkinsfile:184]: We wouldn't want to have a 4th PR build, we would switch one of the existing ones to use Scala 3.0.

**Comment by ijuma** [comment]
Thanks for the PR. Are we sure everything works correctly at the zinc level? Someone mentioned a zinc upgrade here https://github.com/gradle/gradle/issues/16527. The 3.1 release is still a while away, so I would probably wait a bit for the tooling ecosystem to improve for Scala 3.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala (hunk 1 above)
Comment (jlprat) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:793]: This warning made the compilation fail:
```
Usage of named or default arguments transformed this annotation
constructor call into a block. The corresponding AnnotationInfo
will contain references to local values and default getters instead
of the actual argument trees
```

File: Jenkinsfile (hunk 2 above)
Comment (jlprat) [reply Jenkinsfile:184]: Makes sense. Jenkins doesn't pick up the configuration from the PR changes themselves right? On the tasks associated with the PR I didn't see any Scala 3.0 pipeline.

Comment (ijuma) [reply Jenkinsfile:184]: For security reasons, it won't do it until it's merged if the submitter is not a committer.

**Comment by jlprat** [comment]
I agree with you, this can still wait a bit. I just wanted to check if it was a low effort task or a high one.
Regarding if we are sure that everything works correctly, I can successfully compile and run tests, and dependencies are downloaded correctly. I can try to build artifacts and see if it works fine.
Some of the things I needed to play around with in the `dependencies.gradle` file might be redundant if proper Scala 3.0 support is added to Gradle.

**Comment by jlprat** [comment]
Closing this in favour of https://github.com/apache/kafka/pull/11025

File: Jenkinsfile (hunk 2 above)
Comment (jlprat) [reply Jenkinsfile:184]: Makes complete sense. I was assuming this was the behavior.
I wait to change this once the PR is really actionable.

File: gradlewAll [hunk 3]
```diff
@@ -15,5 +15,5 @@
 
 # Convenient way to invoke a gradle command with all Scala versions supported
 # by default
-./gradlew "$@" -PscalaVersion=2.12 && ./gradlew "$@" -PscalaVersion=2.13
+./gradlew "$@" -PscalaVersion=2.12 && ./gradlew "$@" -PscalaVersion=2.13 && ./gradlew "$@" -PscalaVersion=3.0
```
Comment (jlprat) [inline gradlewAll:18]: Just learned of the presence of this file

**Comment by jlprat** [comment]
Something doesn't feel right. I tried to add a piece of Scala 3.0 only syntax on the code to really check it compiles with Scala 3.0 but apparently it fails.
It seems we will need to wait for https://github.com/gradle/gradle/issues/16527 to move forward

**Comment by jlprat** [comment]
I will close this PR as this is not really compiling in Scala3 as Gradle's scala plugin somehow still attempts to compile in Scala 2.

File: gradle/dependencies.gradle [hunk 4]
```diff
@@ -132,7 +145,7 @@ libs += [
   jacksonAnnotations: "com.fasterxml.jackson.core:jackson-annotations:$versions.jackson",
   jacksonDatabind: "com.fasterxml.jackson.core:jackson-databind:$versions.jackson",
   jacksonDataformatCsv: "com.fasterxml.jackson.dataformat:jackson-dataformat-csv:$versions.jackson",
-  jacksonModuleScala: "com.fasterxml.jackson.module:jackson-module-scala_$versions.baseScala:$versions.jackson",
+  jacksonModuleScala: "com.fasterxml.jackson.module:jackson-module-scala_$versions.baseScalaForSpecialLibs:$versions.jackson",
```
Comment (pjfanning) [inline gradle/dependencies.gradle:148]: jackson-module-scala 2.13.0-rc2 is available for scala 3 - generally, older versions of jackson-module-scala will not be able to deal with scala 3 classes - a number of changes have gone into jackson-module-scala 2.13 branch to support scala 3 generated classes correctly
